### PR TITLE
add network-topology-aware design doc

### PR DESCRIPTION
**Pull request by william-wang**
This pr includes a proposal to support the network topology management and network aware scheduling in Volcano. A new network topology CR is designed to define the spine-leaf network in an explicit way. It is also compatible the traditional way using labels.

### CODE REVIEW

File: docs/design/Network Topology Aware Scheduling.md [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+
+In the LLM training scenario, the model parallels have extremely high requirements for the network throughput to exchange data, making the networking to be a bottleneck. There are diverse network in the datacenter, e.g. IB, RoCE, Nvswitch and in each type of network, there might be multiple levels of switch having different throughput and latency. Use requires that the workload can be scheduled to the best performance domain with highest throughput and lowest latency to accelerate data exchanging for training.
+
+Use Case 1
+
+1. A training Job: 8 GPU per Pod \* \~3k (5k) Pods  
+   1. MPI, PyTorch …  
+2. 1 BF3, 8 CX7, 8 H100
+
+![][images/network-topology-aware/network-topology-usecase1]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:15]:
```suggestion
![network topology usecase1][images/network-topology-aware/network-topology-usecase1.png]
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+## 
+
+3. Schedule the job, prefer scheduling all pods to one tier1 topology zone, if not enough nodes, try to schedule all the pods to one tier2 topology zone.  
+4. Gang scheduling is also needed in this case, to make sure all pods are able to proceed with their work. 
+
+## Use Case 2
+
+1. LLM training job:  16 NPU per pod, 3000 pod per job
+
+![][images/network-topology-aware/network-topology-usecase2]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:26]:
```suggestion
![network topology usecase2][images/network-topology-aware/network-topology-usecase2.png]
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+
+2. There are 3 network control planes, which are VPC, Roce and HCCS.  
+3. The pods in Job are expected to be grouped and the pods belonging to the same group have higher demand for the network bandwidth. The tensor parallels are performed on pods in the same group.  
+   ![][images/network-topology-aware/network-topology-usecase3]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:30]:
```suggestion
   ![network topology usecase3][images/network-topology-aware/network-topology-usecase3.png]
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+	……
+}
+
+```
+
+Here are two examples:
+
+Example 1:  
+Here is a spine-leaf tree, there are three levels of switches, connecting the 8 nodes.
+![][images/network-topology-aware/network-topology-example1]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:99]:
```suggestion
![network topology example1][images/network-topology-aware/network-topology-example1.png]
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+   tier: 3
+   members:
+    name: s4
+    name: s5
+
+```
+
+Example 2:  
+There are two roce networks in one Kubernetes cluster which are roce-network0 and roce-network1, and there are two nvlink-networks in each roce network. Each nvlink-network has 4 hosts. The CR is like this.
+![][images/network-topology-aware/network-topology-example2]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:171]:
```suggestion
![network topology example2][images/network-topology-aware/network-topology-example2.png]
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+    name: node7
+    type: Node
+    name: node8
+    type: Node
+```
+
+#### network topology generation and updation
+
+* **Network topology discovery/detection tool**: a tool to generate network topology CR by analyzing labels, system file or API of HW vendor. The community will offer a tool to generate CR by label.  
+![][images/network-topology-aware/nework-topology-generation]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:244]:
```suggestion
![nework topology generation][images/network-topology-aware/nework-topology-generation.png]    
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+                       limits:  
+                          gpu: 8
+```
+
+## Implementation
+
+### Overview
+
+Phase 1 without hyperJob supported  
+![][images/network-topology-aware/network-topology-implementation-01]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:363]:
```suggestion
![network topology implementation 01][images/network-topology-aware/network-topology-implementation-01.png]
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+
+Phase 2 with hyperJob supported  
+![][images/network-topology-aware/network-topology-implementation-02]
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:366]:
```suggestion
![network topology implementation 02][images/network-topology-aware/network-topology-implementation-02.png]  
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 9]
```diff
@@ -0,0 +1,575 @@
+# Network Topology Aware Scheduling
+
+Author: William Wang, Peng Gu, Kevin Wang, Klaus Ma, Xuzheng Chang
+
+# Motivation
+
+In the LLM training scenario, the model parallels have extremely high requirements for the network throughput to exchange data, making the networking to be a bottleneck. There are diverse network in the datacenter, e.g. IB, RoCE, Nvswitch and in each type of network, there might be multiple levels of switch having different throughput and latency. Use requires that the workload can be scheduled to the best performance domain with highest throughput and lowest latency to accelerate data exchanging for training.
+
+Use Case 1
```
Comment (Monokaix) [inline docs/design/Network Topology Aware Scheduling.md:9]: Use Case 1 -> ## Use Case 2

**Comment by Monokaix** [comment]
Please add some description about your pr and what's the difference with https://github.com/volcano-sh/volcano/pull/3552？

File: docs/design/images/network-topology-aware/network-topology-usecase1.png
Comment (JesseStutler) [inline docs/design/images/network-topology-aware/network-topology-usecase1.png:1]: The picture has a saving watermark that needs to be removed.

File: docs/design/Network Topology Aware Scheduling.md (hunk 9 above)
Comment (william-wang) [reply docs/design/Network Topology Aware Scheduling.md]: updated

File: docs/design/Network Topology Aware Scheduling.md (hunk 1 above)
Comment (william-wang) [reply docs/design/Network Topology Aware Scheduling.md]: Updated

File: docs/design/Network Topology Aware Scheduling.md [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+   tier: 1
+   members:
+    name: node0
+    name: node1
+
+version: topology.volcano.sh/v1alpha1
+kind: HyperNode
+name: s1
+spec:
+  HyperNode
```
Comment (JesseStutler) [inline docs/design/Network Topology Aware Scheduling.md:115]: There are redundant `HyperNode` here need to be removed

File: docs/design/Network Topology Aware Scheduling.md [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+   tier: 1
+   members:
+    name: node2
+    name: node3
+
+version: topology.volcano.sh/v1alpha1
+kind: HyperNode
+name: s2
+spec:
+  HyperNode
```
Comment (JesseStutler) [inline docs/design/Network Topology Aware Scheduling.md:125]: Same as above

File: docs/design/Network Topology Aware Scheduling.md [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+type HyperNode struct {
+	metav1.TypeMeta `json:”,inline”`
+	metav1.ObjectMeta `json:”metadata, omitempty”`
+
+	Spec HyperNodeSpec `json:”spec”`
+	Status HyperNodeStatus `json:”status”`
+}
+
+type HyperNodeSpec struct {
+	tier string	`json:"tier,omitempty"`
```
Comment (JesseStutler) [inline docs/design/Network Topology Aware Scheduling.md:80]: tier is a string or int?

File: docs/design/Network Topology Aware Scheduling.md [hunk 13]
```diff
@@ -0,0 +1,575 @@
...
+There are two roce networks in one Kubernetes cluster which are roce-network0 and roce-network1, and there are two nvlink-networks in each roce network. Each nvlink-network has 4 hosts. The CR is like this.
+![network topology example2](images/network-topology-aware/network-topology-example2.png)
+
+```go
+version: topology.volcano.sh/v1alpha1
+kind: HyperNode
+name: roce-network0
+spec:
+tier: 2
+	members:
```
Comment (JesseStutler) [inline docs/design/Network Topology Aware Scheduling.md:179]: The layout here looks a bit messy

File: docs/design/Network Topology Aware Scheduling.md [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+}
+
+type HyperNodeSpec struct {
+	tier string	`json:"tier,omitempty"`
+	members []MemberSpec	`json:"members,omitempty"`
+}
+
+type MemberSpec struct {
+	name string	`json:"name,omitempty"`
+	type string	`json:"type,omitempty"`
```
Comment (JesseStutler) [inline docs/design/Network Topology Aware Scheduling.md:86]: We should capitalize these fields? `type` conflicts with keywords now. Can the `type` field here be enumerated with constants? There are only two types: `Node` and `HyperNode`, right?

File: docs/design/Network Topology Aware Scheduling.md [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+```go
+apiVersion: [batch.volcano.sh/v1aplha1](http://batch.volcano.sh/v1aplha1)  
+kind: HyperJob  
+metadata:  
+    name: multi-volcano-job  
+spec:  
+    networkTopologies: 
+    - mode: hard** // we don’t really need to explicitly indicate the soft requirements, we can make it as a default algorithm behavior. And we need to describe the default behavior in the field comment  
+      highestTierAllowed: 2**  
+    replicatedJobs:
```
Comment (JesseStutler) [inline docs/design/Network Topology Aware Scheduling.md:331]: The fields of this HyperJob sample did not match the API design.

**Comment by hwdef** [comment]
Why create a new hyperJob instead of using an existing vcjob?

I don't think there's anything wrong with the design. What matters is how the topological algorithm is implemented

File: docs/design/Network Topology Aware Scheduling.md [hunk 16]
```diff
@@ -0,0 +1,575 @@
+# Network Topology Aware Scheduling
+
+Author: William Wang, Peng Gu, Kevin Wang, Klaus Ma, Xuzheng Chang
+
+# Motivation
+
+In the LLM training scenario, the model parallels have extremely high requirements for the network throughput to exchange data, making the networking to be a bottleneck. There are diverse network in the datacenter, e.g. IB, RoCE, Nvswitch and in each type of network, there might be multiple levels of switch having different throughput and latency. Use requires that the workload can be scheduled to the best performance domain with highest throughput and lowest latency to accelerate data exchanging for training.
```
Comment (lowang-bh) [inline docs/design/Network Topology Aware Scheduling.md:7]: Use requires -> User requires ?

File: docs/design/Network Topology Aware Scheduling.md [hunk 17]
```diff
@@ -0,0 +1,575 @@
...
+### network topology management
+
+Option 1: Describe network topology by labels, there's a proposal in upstream  
+Cons: 
+* complicated to construct tree for the topology information  
+
+Option 2: Describe network topology by CRD, NetworkTopology in Volcano → plugin/scheduling
+
+- label \-\> NetworkTopology   
+- Rest API (NV, HW) \-\> NetworkTopology
```
Comment (lowang-bh) [inline docs/design/Network Topology Aware Scheduling.md:55]: does the list forget one option of the system file generated by detection tool -> NetworkTopology?

**Comment by lowang-bh** [comment]
I think it is a perfect design.
One question is that if user use label or API to trigger the NetworkTopology CRD, and there is also a auto detection tool,  how the NetworkTopology decide the finall topology info when there are conflicts.

**Comment by Monokaix** [comment]
> I think it is a perfect design. One question is that if user use label or API to trigger the NetworkTopology CRD, and there is also a auto detection tool, how the NetworkTopology decide the finall topology info when there are conflicts.

That's a good question, we just support a way to generate hyperNode crd by default in controller, and if usesr create hyperNode themselves, then they should not use label, that's to say, the two ways conflicts each other, and user should choose  one of them.

**Comment by OguzPastirmaci** [comment]
This is great!

As I mentioned in the other PR, using labels (or having a way to create the NetworkTopology using labels) is a more straightforward way for us initially. For our GPU bare metal nodes, we store the tier 1/2/3 switch info in our metadata service and we label Kubernetes nodes with those info during provisioning. But I'm also very interested in testing out RoCE discovery solution as well to see how/if it would work in a cloud environment with no access to switches. I have test clusters and happy to help with testing.

**Comment by kevin-wangzefeng** [comment]
> > Why create a new hyperJob instead of using an existing vcjob?
> > I don't think there's anything wrong with the design. What matters is how the topological algorithm is implemented
> 
> A hyperJob is list of vcjobs and hyperJob is needed in large model training.

Would be helpful if we can add more description in the design doc to explain the motivation and best practice we'd like to build via HyperJob.

File: docs/design/Network Topology Aware Scheduling.md [hunk 18]
```diff
@@ -0,0 +1,575 @@
...
+![network topology usecase1](images/network-topology-aware/network-topology-usecase1.png)
+
+## 
+
+3. Schedule the job, prefer scheduling all pods to one tier1 topology zone, if not enough nodes, try to schedule all the pods to one tier2 topology zone.  
+4. Gang scheduling is also needed in this case, to make sure all pods are able to proceed with their work. 
+
+## Use Case 2
+
+1. LLM training job:  16 NPU per pod, 3000 pod per job
```
Comment (jxs1211) [inline docs/design/Network Topology Aware Scheduling.md:24]: 3000 pods

**Comment by JesseStutler** [comment]
> This is great!
> 
> As I mentioned in the other PR, using labels (or having a way to create the NetworkTopology using labels) is a more straightforward way for us initially. For our GPU bare metal nodes, we store the tier 1/2/3 switch info in our metadata service and we label Kubernetes nodes with those info during provisioning. But I'm also very interested in testing out RoCE discovery solution as well to see how/if it would work in a cloud environment with no access to switches. I have test clusters and happy to help with testing.

Hi @OguzPastirmaci, how do you label the node? Currently, we disign to label the node such as: `volcano.sh/hypernode:s0,s4,s6", this indicates the tier1/tier2/tier3 switch from near to far that the node can connect to

File: docs/design/Network Topology Aware Scheduling.md [hunk 19]
```diff
@@ -0,0 +1,575 @@
...
+	ExactMatch *ExactMatch `json:"exactMatch,omitempty"`
+
+	// RegexMatch defines the regex match criteria (required when Type is "Regex").
+	// +optional
+	RegexMatch *RegexMatch `json:"regexMatch,omitempty"`
+}
+
+// ExactMatch represents the criteria for exact name matching.
+type ExactMatch struct {
+	// Name specifies the exact name of the node to match.
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:481]:
```suggestion
	// Name specifies the exact name of the node or hyperNode to match.
```

File: docs/design/Network Topology Aware Scheduling.md [hunk 20]
```diff
@@ -0,0 +1,575 @@
...
+	// +optional
+	Name string `json:"name"`
+}
+
+// RegexMatch represents the criteria for regex-based matching.
+type RegexMatch struct {
+	// Pattern defines the regex pattern to match node names.
```
Comment (kevin-wangzefeng) [inline docs/design/Network Topology Aware Scheduling.md:488]:
```suggestion
	// Pattern defines the regex pattern to match node or hyperNode names.
```

**Comment by OguzPastirmaci** [comment]
> > This is great!
> > As I mentioned in the other PR, using labels (or having a way to create the NetworkTopology using labels) is a more straightforward way for us initially. For our GPU bare metal nodes, we store the tier 1/2/3 switch info in our metadata service and we label Kubernetes nodes with those info during provisioning. But I'm also very interested in testing out RoCE discovery solution as well to see how/if it would work in a cloud environment with no access to switches. I have test clusters and happy to help with testing.
> 
> Hi @OguzPastirmaci, how do you label the node? Currently, we disign to label the node such as: `volcano.sh/hypernode:s0,s4,s6", this indicates the tier1/tier2/tier3 switch from near to far that the node can connect to

@JesseStutler here's an example. Local block is tier1, network block is tier2, hpc island is tier3. If using custom labels is not an option, we can relabel the nodes to match the logic needed for the plugin.

```
oci.oraclecloud.com/rdma.local_block_id=abf5v5bcqca
oci.oraclecloud.com/rdma.network_block_id=gbumxwaqp4a
oci.oraclecloud.com/rdma.hpc_island_id=aatoezo236a
```

**Comment by JesseStutler** [comment]
> > > This is great!
> > > As I mentioned in the other PR, using labels (or having a way to create the NetworkTopology using labels) is a more straightforward way for us initially. For our GPU bare metal nodes, we store the tier 1/2/3 switch info in our metadata service and we label Kubernetes nodes with those info during provisioning. But I'm also very interested in testing out RoCE discovery solution as well to see how/if it would work in a cloud environment with no access to switches. I have test clusters and happy to help with testing.
> > 
> > 
> > Hi @OguzPastirmaci, how do you label the node? Currently, we disign to label the node such as: `volcano.sh/hypernode:s0,s4,s6", this indicates the tier1/tier2/tier3 switch from near to far that the node can connect to
> 
> @JesseStutler here's an example. Local block is tier1, network block is tier2, hpc island is tier3. Note that the labels represent groups/tiers of switches and not a single switch (e.g. `rdma.local_block_id` for our GPU rails optimized network will include s0,s1... in it). So the idea is "try to schedule the pods in tier1 nodes only. If not possible, overflow to tier2. If still not enough, overflow to tier3.".
> 
> If using custom labels is not an option, we can try relabeling the nodes to match the logic needed for the plugin.
> 
> ```
> oci.oraclecloud.com/rdma.local_block_id=abf5v5bcqca
> oci.oraclecloud.com/rdma.network_block_id=gbumxwaqp4a
> oci.oraclecloud.com/rdma.hpc_island_id=aatoezo236a
> ```

Thanks! We will discuss again how to unify different labels, and then use volcano's unified label to generate hypernodes. Would you willing to use this feature? Welcome to join our discussion if you want. We can arrange a zoom meeting.

**Comment by OguzPastirmaci** [comment]
> > > > This is great!
> > > > As I mentioned in the other PR, using labels (or having a way to create the NetworkTopology using labels) is a more straightforward way for us initially. For our GPU bare metal nodes, we store the tier 1/2/3 switch info in our metadata service and we label Kubernetes nodes with those info during provisioning. But I'm also very interested in testing out RoCE discovery solution as well to see how/if it would work in a cloud environment with no access to switches. I have test clusters and happy to help with testing.
> > > 
> > > 
> > > Hi @OguzPastirmaci, how do you label the node? Currently, we disign to label the node such as: `volcano.sh/hypernode:s0,s4,s6", this indicates the tier1/tier2/tier3 switch from near to far that the node can connect to
> > 
> > 
> > @JesseStutler here's an example. Local block is tier1, network block is tier2, hpc island is tier3. Note that the labels represent groups/tiers of switches and not a single switch (e.g. `rdma.local_block_id` for our GPU rails optimized network will include s0,s1... in it). So the idea is "try to schedule the pods in tier1 nodes only. If not possible, overflow to tier2. If still not enough, overflow to tier3.".
> > If using custom labels is not an option, we can try relabeling the nodes to match the logic needed for the plugin.
> > ```
> > oci.oraclecloud.com/rdma.local_block_id=abf5v5bcqca
> > oci.oraclecloud.com/rdma.network_block_id=gbumxwaqp4a
> > oci.oraclecloud.com/rdma.hpc_island_id=aatoezo236a
> > ```
> 
> Thanks! We will discuss again how to unify different labels, and then use volcano's unified label to generate hypernodes. Would you willing to use this feature? Welcome to join our discussion if you want. We can arrange a zoom meeting.

Yes, I'm willing to use this feature and happy to join the discussion.

File: docs/design/Network Topology Aware Scheduling.md (hunk 14 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md:86]: Here is just an example, and the API definition is below.

File: docs/design/Network Topology Aware Scheduling.md (hunk 10 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md]: done

File: docs/design/Network Topology Aware Scheduling.md (hunk 12 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md:79]: It's a string.

File: docs/design/Network Topology Aware Scheduling.md (hunk 13 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md]: done

File: docs/design/Network Topology Aware Scheduling.md (hunk 11 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md]: done

File: docs/design/Network Topology Aware Scheduling.md (hunk 15 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md:331]: done

File: docs/design/Network Topology Aware Scheduling.md (hunk 17 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md:53]: Added.

File: docs/design/Network Topology Aware Scheduling.md (hunk 18 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md]: done

File: docs/design/Network Topology Aware Scheduling.md [hunk 21]
```diff
@@ -0,0 +1,575 @@
...
+//	members:
+//	- selector:
+//	    type: Regex
+//	    regexMatch:
+//	      pattern: "^node-[0-9]+$"
+type MemberSelector struct {
+	// Type specifies the selection method (Exact or Regex).
+	// +kubebuilder:validation:Enum=Exact;Regex
+	// // +required
+	Type MemberSelectorType `json:"type"`
```
[the remark this answers is not in the document]
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md]: done

File: docs/design/Network Topology Aware Scheduling.md (hunk 19 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md]: done

File: docs/design/Network Topology Aware Scheduling.md (hunk 20 above)
Comment (Monokaix) [reply docs/design/Network Topology Aware Scheduling.md]: done

**Comment by Monokaix** [comment]
The doc has been updated, please check it once more. @kevin-wangzefeng @hwdef @lowang-bh @JesseStutler

**Comment by JesseStutler** [comment]
Hi all! 👋

Network Topology Aware Scheduling is now available in preview! Check out the docs: https://volcano.sh/en/docs/network_topology_aware_scheduling/

Feel free to try it out and let us know what you think in slack, wechat or github issues 🚀

**Pull request merged**